Add EIP: Elastic Signature (#6327)
ethereum/EIPs: EIPS/eip-6327.md

---
eip: 6327
title: Elastic Signature

> description: Use password to sign data as privatekey

Reviewer: ```suggestion
description: Use password to sign data as private key
```

author: George (@JXRow)
discussions-to: https://ethereum-magicians.org/t/eip-6327-elastic-signature-es/12554
status: Draft
type: Standards Track
category: ERC
created: 2023-01-13
---

## Abstract

> Elastic signature (ES) aims to sign data with a human friendly secret. The secret will be verified fully on-chain and is not stored anywhere. A user can change the secret as often as they need to. The secret does not have a fixed length. The secret will be just like password, which is a better understood concept than private key. This is specifically true for non-technical users. This EIP defines a smart contract interface to verify and authorize operations with ES.

Reviewer: ```suggestion
Elastic signature (ES) aims to sign data with a human friendly secret. The secret will be verified fully on-chain and is not stored anywhere. A user can change the secret as often as they need to. The secret does not have a fixed length. The secret will be like a password, which is a better understood concept than private key. This is specifically true for non-technical users. This EIP defines a smart contract interface to verify and authorize operations with ES.
```

  - SHOULD derive `pwdhash` for a given address. The address can be an EOA or a smart contract wallet.
  - SHOULD verify the proof with the derived `pwdhash`, the computed `fullhash` and a `allhash`, which is submitted by the requester.
- A requester
  - SHOULD generate `datahash` and decide an `expiration`.
  - SHALL request a verification from the verifier with, 
    - `proof` and `allhash` which are provided by the prover;
    - `datahash`;
    - `expiration`.
- A prover
  - SHOULD generate the `proof` and `allhash` from, 
    - `datahash` and `expiration` which are agreed with the requester;
    - `nonce` and other well-known variables. 

There are also some requirements.

- well-known variable SHOULD be available to all parties.
  - SHOULD include a `nonce`.
  - SHOULD include a `chainid`.
  - MAY include any variable that is specific to the verifier.
- public statements SHOULD include, 
  - one reflecting the `pwdhash`;
  - one reflecting the `fullhash`;
  - one reflecting the `allhash`.
- The computation of `fullhash` SHOULD be agreed by both the verifier and the prover.
- The computation of `datahash`

> ### IElasticSignature Interface 

Reviewer: ```suggestion
### `IElasticSignature` Interface 
```

This is the verifier interface.

> ```javascript

Reviewer: ````suggestion
```solidity
````

pragma solidity ^0.8.0;

interface IElasticSignature {
    /**
     * Event emitted after user set/reset their password
     * @param user - an user's address, for whom the password hash is set. It could be a smart contract wallet address
     *  or an EOA wallet address.
     * @param pwdhash - a password hash
     */
    event SetPassword(address indexed user, uint indexed pwdhash);

    /**
     * Event emitted after a successful verification performed for an user
     * @param user - an user's address, for whom the submitted `proof` is verified. It could be a smart contract wallet
     *  address or an EOA wallet address.

>      * @param nonce - a new nonce, which is newly generated to replace the last used nonce. 

Reviewer 2: Does it indicate that `nonce` could be non-sequential?

Author: No, it can't be non-sequential, because the contract using nonce+1 for the next nonce

Reviewer 2: In that case, help me understand why the event needs to emit nonce, if nonce can be inferred?

## Rationale

The contract will store everyone's `pwdhash`.

![verifier-contract](../assets/eip-6327/zkpass-1.png)

The chart below shows ZK circuit logic.

![circuit-logic](../assets/eip-6327/zkpass-2.png)

To verify the signature, it needs `proof`, `allhash`, `pwdhash` and `fullhash`.

![workflow](../assets/eip-6327/zkpass-3.png)

The prover generates `proof` along with the public outputs. They will send all of them to a third-party requester contract. The requester will generate the `datahash`. It sends `datahash`, `proof`, `allhash`, `expiration` and prover's address to the verifier contract. The contract verifies that the `datahash` is from the prover, which means the withdrawal operation is signed by the prover's password.

## Backwards Compatibility

This EIP is backward compatible with previous work on signature validation since this method is specific to password based signatures and not EOA signatures. 

## Reference Implementation

Example implementation of a signing contract:

> ```javascript

Reviewer: ````suggestion
```solidity
````

        emit Verified(user, nonce);
    }

    /////////// util ////////////

    function verifyProof(
        uint[8] memory proof,
        uint pwdhash,
        uint fullhash, //254b
        uint allhash
    ) internal view returns (bool) {
        return
            verifier.verifyProof(
                [proof[0], proof[1]],
                [[proof[2], proof[3]], [proof[4], proof[5]]],
                [proof[6], proof[7]],
                [pwdhash, fullhash, allhash]
            );
    }
}
```

verifier.sol is auto generated by snarkjs, the source code circuit.circom is below

```javascript
pragma circom 2.0.0;

include "../../node_modules/circomlib/circuits/poseidon.circom";

template Main() {
    signal input in[3];
    signal output out[3];

    component poseidon1 = Poseidon(2);
    component poseidon2 = Poseidon(2);

    poseidon1.inputs[0] <== in[0];  //pwd
    poseidon1.inputs[1] <== in[1];  //address
    out[0] <== poseidon1.out; //pwdhash

    poseidon2.inputs[0] <== poseidon1.out;
    poseidon2.inputs[1] <== in[2]; //fullhash
    out[1] <== in[2]; //fullhash
    out[2] <== poseidon2.out; //allhash
}

component main = Main();
```

## Security Considerations

> Since the pwdhash is public, it is possible to be crack the password. We estimate the Poseidon hash rate of RTX3090 would be 100Mhash/s, this is the estimate of crack time:

Reviewer 2: The EIP looks good as a draft. And thank you for this creative idea!

[Peer Review] The only technical concern I have is that in 2008-2009 the Bitcoin's hashrate globally is 10MHash/s, today it's 400EHash/s which is 4e13 times more. If the hash rate of Poseidon increases as fast in that rate, it will be cracked in 100seconds in 15th year.  116586246y = 3.6766638539e+15sec

Author: Thank you, thanks for the notice, we are following, if any better way, we will update
